### DISCUSSION TITLE
mutiple subscribers in a single script

### DISCUSSION THREAD

**Discussion by NehalNevle** [eclipse-iceoryx/iceoryx#1768] [Q&A]
Is it possible to have multiple subscribers in a single script getting data from different publisher (one to one)?
![WhatsApp Image 2022-10-29 at 16 05 24](https://user-images.githubusercontent.com/107995795/198826708-29da64c2-cbff-4e17-b1cb-4ddda6336fa2.jpeg)

**Comment by NehalNevle** [comment]
@rotu @lygstate @clalancette @elfenpiff

**Reply by elfenpiff** [reply]
@NehalNevle Yes this is possible. Just create two subscribers and work with them as it is described in the icehello example: https://github.com/eclipse-iceoryx/iceoryx/tree/master/iceoryx_examples/icehello

For instance, based on the icehello example:
```cxx
iox::popo::Subscriber<RadarObject> subscriber({"Radar", "FrontLeft", "Object"});
iox::popo::Subscriber<RadarObject2> subscriber2({"Some", "Other", "Service"});

while(true) {
  auto takeResult = subscriber.take();
  auto takeResult2 = subscriber2.take();

  if ( !takeResult.has_error() ) { std::cout << takeResult.value()->x << std::endl; }
  if ( !takeResult2.has_error() ) { std::cout << takeResult.value()->x << std::endl; }
}
```

Usually you do not want to do this in a busy loop since this costs you a lot of CPU time. If you handle multiple subscribers and want to be triggered only when a subscriber has received something I suggest to use the waitset, see this example: https://github.com/eclipse-iceoryx/iceoryx/tree/master/iceoryx_examples/waitset
and the file: https://github.com/eclipse-iceoryx/iceoryx/blob/master/iceoryx_examples/waitset/ice_waitset_gateway.cpp

Applied to this little example you only have to add this code:
```cxx
iox::popo::WaitSet<> waitset;
waitset.attachEvent(subscriber, iox::popo::SubscriberEvent::DATA_RECEIVED)
            .or_else([&](auto) {
                std::cerr << "failed to attach subscriber" << i << std::endl;
                std::exit(EXIT_FAILURE);
            });
waitset.attachEvent(subscriber2, iox::popo::SubscriberEvent::DATA_RECEIVED)
            .or_else([&](auto) {
                std::cerr << "failed to attach subscriber" << i << std::endl;
                std::exit(EXIT_FAILURE);
            });

while(true) {
  auto notificationVector = waitset.wait();
  for (auto& notification : notificationVector) {
    if ( notification->doesOriginateFrom(&subscriber) { /* print subscriber message */ }
    else if ( notification->doesOriginateFrom(&subscriber2) { /* print subscriber message */ }
  }
}
```
